base/game_events/PsynergyStoneEvent.ts

```diff
@@ -0,0 +1,154 @@
+import {CharsStatusWindow} from "../windows/CharsStatusWindow";
+import {NPC} from "../NPC";
+import {base_actions, directions} from "../utils";
+import {DialogManager} from "../utils/DialogManager";
+import {GameEvent, event_types} from "./GameEvent";
+import {Button} from "../XGamepad";
+import * as _ from "lodash";
+
+const INIT_TEXT = (hero_name: string) => `${hero_name} checked on the ground...`;
+const GET_TEXT = (hero_name: string) => `${hero_name} got a Psynergy Stone.`;
+const RECOVERY_TEXT = (hero_name: string) => `${hero_name}'s PP are maxed out!`;
+const DISAPPEAR_TEXT = `The Psynergy Stone disappeared...`;
+
+export class PsynergyStoneEvent extends GameEvent {
+    private control_enable: boolean = true;
+    private running: boolean = false;
+    private resolve: Function;
+    private promise: Promise<any>;
+    private dialog_manager: DialogManager;
+    private custom_init_text: string;
+    private chars_status_window: CharsStatusWindow;
+    private finish_events: GameEvent[] = [];
+
+    constructor(game, data, active, finish_events, custom_init_text, hide_on_finish) {
+        super(game, data, event_types.STONE, active);
+        this.custom_init_text = custom_init_text;
+        this.chars_status_window = new CharsStatusWindow(this.game, this.data);
+
+        this.data.control_manager.add_controls(
+            [
+                {
+                    button: Button.A,
+                    on_down: () => {
+                        if (!this.running || !this.control_enable) return;
+                        this.next();
+                    },
+                },
+            ],
+            {persist: true}
+        );
+
+        finish_events?.forEach(event_info => {
+            const event = this.data.game_event_manager.get_event_instance(event_info);
+            this.finish_events.push(event);
+        });
+    }
+
+    next() {
+        this.control_enable = false;
+        this.dialog_manager.next(finished => {
+            this.control_enable = true;
+            if (finished) {
+                this.resolve();
+            }
+        });
+    }
+
+    async _fire(origin_npc?: NPC) {
+        if (!this.active) return;
+        this.origin_npc = origin_npc;
+        ++this.data.game_event_manager.events_running_count;
+        this.control_enable = false;
+        this.running = true;
+
+        //Initial Text
+        const hero_name = this.data.info.party_data.members[0].name;
```

Comment:
Please use this:
`const hero_name = this.data.info.main_chars[this.data.hero.key_name].name;`
In the future, I plan to have other chars as the controllable hero without necessarily putting him at the 0 index.